Testing/SystemTests/lib/systemtests/systemtesting.py

```diff
@@ -499,6 +499,20 @@ def mismatchWorkspaceName(reference_filename):
         name = reference_filename.split('.')[0]
         return name + '-mismatch.nxs'
 
+#########################################################################
+# Add setUp/tearDown methods to create/remove a temporary working
+# directory
+#########################################################################
+class TemporaryWorkingDirectory:
```

Comment:
I just had a quick peek because of the high priority tag (so this isn't me stealing Sarah's review)

As the current pattern will leak tmp directories and files if it throws or dies in an unexpected way, could we do this as a class decorator or metaclass to re-jig this into:
```
try:
    self._work_dir = tempfile.mkdtemp()
    yield cls
finally:
    shutil.rmtree(self._work_dir)
```

As we tend to find the Windows / ORNL RHEL builders run out of disk space with un-tidied cruft

Reply:
I was tempted by that but went this way for simplicity so that the directory only gets created when the test executes. The whole execution is in a `try...finally` (https://github.com/mantidproject/mantid/blob/36a359a6ee7e68421923503e33344a74690df424/Testing/SystemTests/lib/systemtests/systemtesting.py#L229) anyway so `tearDown` will run regardless...

Reply:
Sorry I was definitely not at my best yesterday. We don't override `tearDown` in the system test base class but we do ensure it is called so I don't think leaking is a real problem but saying that, with a fresh pair of eyes I really like the approach of the first link as it means you only need to write normal unittest code with `setUp`. Also the directory handling is not off somewhere else that can be removed accidentally by definining `setUp` on the main class. I'll make those changes.

Reply:
To be fair, I was about to comment that `tearDown` was the way to go until I stumbled onto that link too, so it was just a fortuitous last minute find